### PR TITLE
XHR Monitor fixes, v3 (bare-bones)

### LINKED ISSUE #214: XHR Monitor bugs
**Describe the bug**
The script install_monitor.js sometimes gets stuck, thinking there are unresolved pending XHR requests which in actuality have been garbage-collected or aborted.  This can lead to unnecessary delays in QWeb keywords which wait (upto a timeout) for the `window.__xhrMon.pending` count to reach 0 before proceeding.

I've dug into the XHR problem, and identified two root causes which are closely related to each other:
1) XHR requests which are sent (and reach readyState = 1 immediately after sending), but are garbage-collected before completing.  Presumably, these requests were abandoned by the code which created them, leaving no strong references remaining in the global scope.
2) XHR requests which are sent (and reached readyState = 1 immediately after sending), but inexplicably revert to readyState = 0 without ever invoking the "loadend" event listener.

In both cases, the "loadend" event listener is never invoked, and consequently the `window.__xhrMon.pending` number is never decremented, leaving the XHR Monitor hanging, believing there are still unresolved pending XHR request(s).

Based on my experiments, it seems an XHR request which may be subject to one of the above scenarios can also be prone to the other.  It is simply a question of whether the JS engine's garbage-collector gets to it first, or the readyState reverts to 0 first.

**To Reproduce**
1. Enable debugging in the install_monitor.js script, (on line 6, set `debug = true`)
2. Start a robotic script in debug mode, which opens the web browser, navigates to www.google.com, and calls some QWeb keyword to install the XHR Monitor (e.g. `Verify Text      About`)
3. Open the browser dev console to observe the debug logs
4. Return to the browser and start typing in the search box
5. With each keystroke a new XHR is sent out.  However, if you press multiple keys in rapid succession, a new XHR is sent before the previous ones returns.  There is something about Google's client-server interaction which causes the old XHR to be dropped without ever triggering the "loadend" event listener.  This causes the `window.__xhrMon.pending` to never return to 0.

**Expected behavior**
The `window.__xhrMon.pending` value should accurately reflect the number of outstanding XHR requests still pending.

**Desktop (please complete the following information):**
 - OS: Windows 11
 - Browser Chrome 148.0.7778.97
 - Chromedriver version: 148.0.778.167
 - QWeb Version 3.7.4

**Additional context**

I have already written a patch to this problem, which I will submit as a pull request soon.  Additionally, there a few other minor issues I identified in the XHR Monitor, which I will fix in my PR.

1) The monitor is not installed on the webpage until a QWeb keyword is explicitly called.  Any XHR sent prior to that point will be missed.  For chromium-based browsers, we can use a CDP command (via Selenium webdriver) to install a script that is evaluated on new documents before the DOM loads: `driver.execute_cdp_cmd("Page.addScriptToEvaluateOnNewDocument", {"source": JS_INSTALL_MONITOR})`  Note: the one downside to this is that `document.documentElement` is null at the time this script is evaluated.  The DOM mutation monitor normally observes the `document.documentElement`.  Alternatively, the monitor can listen to just `document` (which does exist at that point) without impacting results.
2) Related to 1., it is common for a webpage's JS to monkey-patch native methods of in-built types and functions which the XHR monitor depends on.  This could alter the XHR Monitor's behavior in a detrimental way, especially if the monitor is installed after the website's monkey-patching is already in-place.  However, if the monitoring script is installed to evaluate on new document (using the CDP command mentioned in 1.), then it can access and store the original native methods before any monkey-patching is done.
3) The DOM monitor's `window._xhrMon.lastMutationTs` timestamp relies upon `Date.now()` which is not a monotonic clock, and could be altered by a change in the system time.  I changed all `Date.now()` calls to `performance.now()` which is monotonic.  This effects "install_monitor.js" and "wait_status_lite.js"

### PR DESCRIPTION
**Pull request by billyJoePiano**
This is version 3 of my proposed fix to the XHR Monitor, based on previous discussions in Pull Requests:
https://github.com/qentinelqi/qweb/pull/218
https://github.com/qentinelqi/qweb/pull/216

It solves the bug described in issue https://github.com/qentinelqi/qweb/issues/214

All unit-tests and acceptance-tests passed on my local system (Windows 11, Python 3.12.10)

This v3 proposal is the most bare-bones possible fix I'd recommend using.  Anything less and there would be considerable edge-case scenarios where the XHR Monitor could break (see also: https://github.com/qentinelqi/qweb/pull/218#issuecomment-4545678323 )

### CODE REVIEW

**Comment by tkoukkari** [comment]
@billyJoePiano Thanks, this is much closer to the kind of narrow-scope fix I had in mind.

I like that this keeps the change limited to the JS monitor logic and avoids the broader browser-startup / CDP changes from the earlier version. On a quick review this looks much more in line with what I’d be comfortable merging.

I still want to do a bit more testing on my side before I approve it, mainly just to make sure it behaves well in practice and that we are not introducing any new wait-related edge cases.One small thing I noticed: if the same XMLHttpRequest object is reused across multiple open() / send() cycles, I think the current __xhrSent / __xhrDone flags may carry over from the first request. In that case, the later request might not get tracked correctly. It seems like the simplest fix would be to reset those per-request flags in open(), while still attaching the listeners only once.

Overall though, this looks much closer to the scope I was hoping for. I’m planning to close the previous two PRs related to this issue, and treat this one as the replacement PR that we continue evaluating.

File: QWeb/internal/js/install_monitor.js [hunk 1]
```diff
@@ -11,67 +11,109 @@ return (function (debug = false) {
...
+					if (debug) {
+						console.log("XHR monitor: request ended, pending count:", window.__xhrMon.pending);
+					}
+				} catch(e) {
+					if (debug) console.error("XHR monitor: error in done", e);
+				}
+			}
+			function readystatechange() {
+				const readyState = _getReadyState.apply(this);
+				if (readyState === 4 || readyState === 0) {
```
Comment (turunenm) [inline QWeb/internal/js/install_monitor.js:110]: Add a comment what these `readyState`s are and/or add link to documentation where the states are described.
Ideally these should be checked against an enum but given the nature of the monitor a comment will work just fine.

**Comment by billyJoePiano** [comment]
> One small thing I noticed: if the same XMLHttpRequest object is reused across multiple open() / send() cycles, I think the current __xhrSent / __xhrDone flags may carry over from the first request. In that case, the later request might not get tracked correctly. It seems like the simplest fix would be to reset those per-request flags in open(), while still attaching the listeners only once.

Interesting.  I didn't realize an XHR could be reused, but I looked it up and you are correct.  It's a non-standard usage, but it can be done.  I'll adjust the code accordingly.

> Overall though, this looks much closer to the scope I was hoping for. I’m planning to close the previous two PRs related to this issue, and treat this one as the replacement PR that we continue evaluating.

Thanks

**Comment by billyJoePiano** [comment]
Hi @tkoukkari -- I just added a few lines in the JS to deal with the XHR re-use scenario.  I noticed that the checks have already started, so you may want to re-run the checks with the new code.

**Review by tkoukkari — APPROVED** [review]
Thanks for iterating on this and for working through the edge cases. This latest version looks good to me.

As far as I can tell, and to the best of my ability to test this so far, everything now looks good and performs as expected. I like that the fix stays narrowly scoped to the JS monitor logic and avoids the broader browser-startup / CDP changes from the earlier versions. The reused-XHRs concern also looks addressed now.

I ran additional testing on my side, including the cases we discussed, and it behaved correctly in those tests as well as the other checks I ran.

Approved from my side.

Thanks again for digging into this and for sticking with the iterations here. This was a very helpful contribution!

**Comment by billyJoePiano** [comment]
> Thanks again for digging into this and for sticking with the iterations here. This was a very helpful contribution!

Thank you!  Glad to contribute!  I hope this contribution proves useful.

**Review by turunenm — APPROVED** [review]
